packages/undo-manager/src/types.ts

```diff
@@ -0,0 +1,19 @@
+export type HistoryChange = {
+	from: any;
+	to: any;
+};
+
+export type HistoryChanges = {
+	id: string | Record< string, any >;
+	changes: Record< string, HistoryChange >;
+};
+
+export type HistoryRecord = Array< HistoryChanges >;
+
+export type UndoManager = {
```

Comment:
Is it possible to properly type the kind of values that the `UndoManager` manages? It's certainly complicated, and I'm wondering whether it's worth it. The value types can be completely different for different `id`s etc.

Reply:
By the way, it would be nice to write the undo manager in TypeScript from the very beginning.

Reply:
I personally prefer TS + JSDocs over typescript but you know at least it's typed 100%.

For the values types, the undo manager is agnostic of the values themselves, so no, we can't type them. Maybe generics but even there it's very hard as each property has its own type basically.